python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,505 @@
+        self.model = model
+        self.order = get_tag_from_model_func(func=model,
+                                             tag='preferred_input_order',
+                                             default='C')
+        if gpu_model is None:
+            # todo: when sparse support is added, use this tag to see if
+            # model can accept sparse data
+            self.model_gpu_based = \
+                get_tag_from_model_func(func=model,
+                                        tag='X_types_gpu',
+                                        default=False) is not None
+        else:
+            self.model_gpu_based = gpu_model
+
+        if output_type is None:
+            self.output_type = 'cupy' if self.model_gpu_based else 'numpy'
+        else:
+            self.output_type = output_type
+
+        # if not dtype is specified, we try to get it from the model
+        if dtype is None:
+            self.dtype = get_dtype_from_model_func(func=model,
+                                                   default=np.float32)
+        else:
+            self.dtype = np.dtype(dtype)
+
+        self.background, self.N, self.M, _ = \
+            input_to_cupy_array(data, order='C',
+                                convert_to_dtype=self.dtype)
+
+        self.nsamples = 2 * self.M + 2 ** 12 if nsamples is None else nsamples
+
+        self.max_samples = 2 ** 30
+
+        # restricting maximum number of samples for memory and performance
+        # value being checked, right now based on mainline SHAP package
+        self.max_samples = 2 ** 30
+        if self.M <= 30:
+            self.max_samples = 2 ** self.M - 2
+            if self.nsamples > self.max_samples:
+                self.nsamples = self.max_samples
+
+        if isinstance(data, pd_df) or isinstance(data, cu_df):
+            self.feature_names = data.columns.to_list()
+        else:
+            self.feature_names = [None for _ in range(len(data))]
+
+        cur_nsamples = 0
+        self.nsamples_exact = 0
+        r = 0
+
+        # we check how many subsets of the powerset of self.M we can fit
+        # in self.nsamples
+        while cur_nsamples <= self.nsamples:
+            r += 1
+            self.nsamples_exact = cur_nsamples
+            cur_nsamples += int(binomCoef(self.M, r))
+
+        # see if we need to have randomly sampled entries in our mask
+        # and combinations matrices
+        self.nsamples_random = \
+            self.nsamples - self.nsamples_exact if r < self.M else 0
+
+        # we save r so we can generate random samples later
+        self.randind = r
+
+        # using numpy powerset and calculations for initial version
+        # cost is incurred only once, and generally we only generate
+        # very few samples if M is big.
+        mat, weight = powerset(self.M, r - 1, self.nsamples_exact,
+                               dtype=self.dtype)
+
+        self.mask = cp.zeros((self.nsamples, self.M), dtype=np.float32)
```

Comment:
If you are creating cupy arrays manually, this needs to be wrapped in a `@cuml.internals.api_return_any()` or at the very least `@with_cupy_rmm`